h2/stream.py

```diff
@@ -735,6 +737,9 @@ def send_headers(self, headers, encoder, end_stream=False):
         except AttributeError:
             headers = headers
 
+        # store request method for _initialize_content_length
+        self.request_method = extract_method_header(headers)
+
```

Comment:
So it may be worth moving this towards the bottom of this method. That allows us to check `self.state_machine.client` and `input_` to work out whether or not we're likely to find a method, and so we can avoid doing unnecessary searches of the header block.